### PR TITLE
Data loading and training loop implementation with lightning

### PR DESCRIPTION
**Pull request by ziw-liu**
The data loading module uses a thin PyTorch `Dataset` class for sampling and a `LightningDataModule` to define transforms for different stages in the pipeline.

The training loop is implemented with a `LightningModule` whose `forward()` method simply wraps an existing U-Net implementation. The trainer is defined from the CLI and config file which are automatically generated.

### CODE REVIEW

File: micro_dl/light/data.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+        z = tz - t * zs
+        selection = (int(t), int(ch_idx), slice(z, z + self.z_window_size))
+        data = img[selection][np.newaxis]
+        return torch.from_numpy(data)
+
+    def __len__(self) -> int:
+        return self._max_window
+
+    def __getitem__(self, index: int) -> dict[str, torch.Tensor]:
+        w, tz = self._find_window(index)
```
Comment (Christianfoley) [inline micro_dl/light/data.py:81]: If I'm understanding this correctly this means we would sample one "valid" patch from each position each epoch (assuming batch size of 1)?

File: micro_dl/light/data.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+            self._read_img_window(img, self.target_ch_idx, tz)
+        )
+        sample = {"source": source, "target": target}
+        if self.transform:
+            sample = self.transform(sample)
+        if isinstance(sample, list):
+            sample = sample[0]
+        if self.target_center_slice_only:
+            sample["target"] = sample["target"][:, self.z_window_size // 2][:, None]
+        return sample
```
Comment (Christianfoley) [inline micro_dl/light/data.py:96]: Does pytorch lightning handle transferring samples to and from GPU?

**Review by Christianfoley — COMMENTED** [review]
This looks very good. I had no idea pytorch lightning was this powerful, especially in combination with iohub and monai. I left some questions in comments.

File: micro_dl/light/data.py (hunk 2 above)
Comment (ziw-liu) [reply micro_dl/light/data.py:96]: Yes it does.

File: micro_dl/light/data.py (hunk 1 above)
Comment (ziw-liu) [reply micro_dl/light/data.py:81]: Yes, and it's also true when the batch size is larger than 1.

**Comment by mattersoflight** [comment]
Looks great to me from tensor board logs and reading the code! I am merging and will continue testing and reviewing.

**Pull request merged**